src/eko/anomalous_dimensions/__init__.py

```diff
@@ -60,22 +60,54 @@ def exp_singlet(gamma_S):
     """
     # compute eigenvalues
     det = np.sqrt(
-        np.power(gamma_S[0, 0] - gamma_S[1, 1], 2) + 4.0 * gamma_S[0, 1] * gamma_S[1, 0]
+        np.power(gamma[0, 0] - gamma[1, 1], 2) + 4.0 * gamma[0, 1] * gamma[1, 0]
     )
-    lambda_p = 1.0 / 2.0 * (gamma_S[0, 0] + gamma_S[1, 1] + det)
-    lambda_m = 1.0 / 2.0 * (gamma_S[0, 0] + gamma_S[1, 1] - det)
+    lambda_p = 1.0 / 2.0 * (gamma[0, 0] + gamma[1, 1] + det)
+    lambda_m = 1.0 / 2.0 * (gamma[0, 0] + gamma[1, 1] - det)
     # compute projectors
     identity = np.identity(2, np.complex_)
     c = 1.0 / det
-    e_p = +c * (gamma_S - lambda_m * identity)
-    e_m = -c * (gamma_S - lambda_p * identity)
+    e_p = +c * (gamma - lambda_m * identity)
+    e_m = -c * (gamma - lambda_p * identity)
     exp = e_m * np.exp(lambda_m) + e_p * np.exp(lambda_p)
     return exp, lambda_p, lambda_m, e_p, e_m
 
 
+@nb.njit(cache=True)
+def exp_matrix(gamma):
+    r"""
+    Compute the exponential and the eigensystem of a matrix.
+
+    Parameters
+    ----------
+        gamma : numpy.ndarray
+            input matrix
+
+    Returns
+    -------
+        exp : numpy.ndarray
+            exponential of the matrix gamma :math:`\gamma(N)`
+        w : numpy.ndarray
+            array of the eigenvalues of the matrix lambda
+        e : numpy.ndarray
+            projectors on the eigenspaces of the matrix gamma :math:`\gamma(N)`
+    """
+    # compute Matrix of coefficients
+    w, v = np.linalg.eig(gamma)
+    tmp = np.linalg.inv(v)
+    dim = gamma.shape[0]
+    e = np.zeros((dim, dim, dim), np.complex_)
+    exp = np.zeros((dim, dim), np.complex_)
+    # TODO check if this loop can be entirely cast to numpy
```

Comment:
maybe @AleCandido can double check, but I'm afraid the answer is no. Looking to [`outer`](https://numpy.org/doc/stable/reference/generated/numpy.outer.html?highlight=outer#numpy.outer) there is no `axis` or similar and `einsum` which is given as an alternative option [is not in numba](http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html#linear-algebra) ... actually this is maybe even not needed since we're actually inside numba and so a loop should not be too expensive

Reply:
The second line should be simple to do it with NumPy
```py
    for i in range(dim):
        e[i] = np.outer(v[:, i], tmp[i])

     exp = (e * np.exp(w)).sum(axis=0)
    return exp, w, e
```

Reply:
> (not 100% sure about the axis, but I almost sure you are summing over the most external one, so 0)

And a trivial test at least shows it working

Reply:
For `e` if I understood the einsum would be
```py
e = np.einsum("ij,jk->jik", v, tmp)
```
so the full one would be:
```py
exp = np.einsum("ij,jk,j->ik", v, tmp, np.exp(w))
```

Reply:
@AleCandido numba doesn't like it: ```Use of unsupported NumPy function 'numpy.einsum' or unsupported use of the function.```

Reply:
This should be a solution, but it is not optimal:
```py
e = np.outer(v, tmp).reshape(dim, dim, dim, dim)[:, idx, idx, :]
exp = (e * np.exp(w)).sum(axis=0)
```

Reply:
> @AleCandido numba doesn't like it: `Use of unsupported NumPy function 'numpy.einsum' or unsupported use of the function.`

I know, @felixhekhorn wrote it above. I just wanted to confirm it is the correct signature, and I used that to write Einstein notation without latex, nothing else...

Reply:
If you want test the proposed solution, but profile it. I expect to be slower than yours, since it requires more operations (and more memory, but most likely it is irrelevant), and Numba should convert the loop to the same loops of NumPy.

If it is not slower, than it's a problem, i.e. Numba it is not doing enough.

Reply:
It still can't be compiled by numba. Moreover in this way both ```e``` and ```exp``` are different with respect to the original function. I would return to the ```for``` loop.

Reply:
Ok, in my test they were the same, but we can check later on. How did you check it if it didn't compile?
In any case, it was more a test to benchmark the original solution, that I believe to be performant enough.

Only for my curiosity, can you tell me what Numba is not able to compile?

Reply:
I run the test ```test_ad.py``` where this function is compared with ```scipy.linalg.expm``` and it is broken.

Regarding numba, it stops at ```e = np.outer(v, tmp).reshape(dim, dim, dim, dim)[:, idx, idx, :]``` with the error
```
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function getitem>) found for signature:
 
 >>> getitem(array(complex128, 4d, C), Tuple(array(int64, 1d, C), slice<a:b>, array(int64, 1d, C), slice<a:b>))
```

Reply:
Ah, `getitem` are the square brackets. This was unexpected...

Reply:
Since you are testing them in `test_ad.py`, you can (also in the same test function) benchmark the performances of the two implementations with [`timeit`](https://docs.python.org/3/library/timeit.html) (standard library module).
You can also assert that one has not to be much worse than the other :)

Reply:
Pass a lambda to `timeit.timeit`:
```py
scipy_time = timeit.timeit(lambda: scipy.linalg.expm(mymat))
numba_time = timeit.timeit(lambda: exp_matrix(mymat))
assert numba_time < 2*scipy_time
```
Where 2 is random value, pick your favorite...
(you can use also `timeit.repeat` to receive a bunch of attempts, and compute mean and variance for the comparison)

Reply:
if this is resolved - please remove the TODO